src/loader-new.js

```diff
@@ -0,0 +1,342 @@
+      </circle>
+      <circle class="i-amphtml-new-loader-spinner-segment" cx="60" cy="60">
+      </circle>
+    </g>`;
+
+    this.svgRoot_.appendChild(spinner);
+  }
+
+  /**
+   * Adds the default or branded logo.
+   * @private
+   */
+  maybeAddLogo_() {
+    const logo = this.getLogo_();
+
+    // Ads always get the logo regardless of size
+    if (this.isAd_()) {
+      return this.addLogo_(logo);
+    }
+
+    // Small hosts do not get a logo
+    if (this.isSmall_()) {
+      return;
+    }
+
+    return this.addLogo_(logo);
+  }
+
+  /**
+   * Returns the logo for the element.
+   * @private
+   * @return {!Element}
+   */
+  getLogo_() {
+    const logo = this.getCustomLogo_();
+    return logo || this.getDefaultLogo_();
+  }
+
+  /**
+   * Returns the custom logo for the element if there is one.
+   * @private
+   * @return {?Element}
+   */
+  getCustomLogo_() {
+    // Not Implemented
+    return null;
+  }
+
+  /**
+   * Returns the default logo.
+   * @private
+   * @return {!Element}
+   */
+  getDefaultLogo_() {
+    return svgFor(this.doc_)`<circle class="i-amphtml-new-loader-logo"
+      cx="60" cy="60" r="12" fill="#aaaaaa">
+    </circle>`;
+  }
+
+  /**
+   * Adds the given logo to the loader.
+   * @param {!Element} logo
+   */
+  addLogo_(logo) {
+    this.svgRoot_.appendChild(logo);
+  }
+
+  /**
+   * Adds the background shim under the loader for cases where loader is on
+   * top of an image.
+   * @private
+   */
+  maybeAddBackgroundShim_() {
+    if (!this.hasImagePlaceholder_()) {
+      return;
+    }
+
+    // Not Implemented
+
+    return;
+  }
+
+  /**
+   * Add a gray default placeholder if there isn't a placeholder already and
+   * other special cases.
+   * @private
+   */
+  maybeAddDefaultPlaceholder_() {
+    // NOTE(aghassemi): I do believe we need to exclude amp-list here, but
+    // let's see how experimentation goes. Maybe a better idea is to have a
+    // white list, any component that does not fully load a new background may
+    // look bad if a gray placeholder shows up and goes away quickly. This
+    // default placeholder is good for image, video, etc.. but amp-list which
+    // usually just loads text is debatable. amp-iframe is also a candidate
+    // to exclude, often it does load a video or maps but it may not load
+    // text on transparent background in certain cases.
+    if (!hasPlaceholder(this.element_)) {
+      this.loadingContainer_.classList.add('i-amphtml-default-placeholder');
+    }
+  }
+
+  /**
+   * Whether the element is an Ad.
+   * @private
+   * @return {boolean}
+   */
+  isAd_() {
+    // Not Implemented
+    return false;
+  }
+
+  /**
+   * Whether the element is small.
+   * Small elements get a different loader with does not have a logo and is
+   * just a spinner.
+   * @private
+   * @return {boolean}
+   */
+  isSmall_() {
+    const box = this.element_.getLayoutBox();
+    return !isTiny(this.element_) && (box.width <= 100 || box.height <= 100);
```

Comment:
Nit: Would be nice to avoid an extra `getLayoutBox()` call which has a small cost. Or we can add `getLayoutSize()`.

Reply:
N/A with the new approach (cached layout as instance var)